### PR TITLE
[ASR Examples] Only filter training samples by audio length criterion

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

We should only filter training samples by our audio length criterion when fine-tuning ASR systems. The eval and test sets should **not** be filtered. Doing so leads to partial datasets, and thus in-valid results.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten @anton-l

### CODE REVIEW

**Review by anton-l — APPROVED** [review]
Good call @sanchit-gandhi! Although some datasets like Libri might lead to OOMs now during eval, so we could add a separate eval duration cutoff as well

**Comment by sanchit-gandhi** [comment]
I'm reluctant to add a cut-off for eval: in filtering the dataset by an audio length criterion, one removes the longest samples in the eval set. Evaluating on a partial eval set yields WER metrics that are not valid for academic purposes. We need to have the full eval set in order to compare WER figures between models. You can't compare a WER for a partial sub-set A with the WER for a different partial sub-set B!

Take the extreme example where you have a model that performs well on short samples <= 5s, and terribly for longer samples > 5s. Filtering with a cut-off of 5s would hide the fact the model breaks down for longer samples. You couldn't then compare this WER metric with another model where you've not filtered the eval samples, and thus included all the samples (including the 'harder' ones > 5s).

IMO, if not filtering eval samples poses an OOM threat, it has to be addressed by a reduction in batch size (chosen by the user depending on their model/hardware/dataset). WDYT @anton-l @patrickvonplaten?

**Comment by patrickvonplaten** [comment]
More or less agree here - just in terms of backwards compatibility, I think this might lead to a couple of unexpected errors. Max length can prevent OOM and min length can prevent weird Conv layer doesn't have enough input values. 

So I'd maybe prefer to instead add a `test_min_duration_in_seconds` and `test_max_duration_in_seconds` here, let it default to `min_duration_in_seconds` and `max_duration_in_seconds` (e.g. set to None and if None, set to `min_duration_in_seconds`) and add a warning that this will be changed in the next versions.

**Comment by patrickvonplaten** [comment]
This fine-tuning script is actually used quite a bit and usually the cutting doesn't make a huge difference (also note in the examples, we're not looking for the "perfect" script, but rather an example of how ctc training could be done)

But overall ok with the PR, just think we should make it more backwards compatible

**Comment by patrickvonplaten** [comment]
@sanchit-gandhi, let me know once you want to merge this :-)

**Comment by sanchit-gandhi** [comment]
Failing test unrelated - rebasing should fix this. Will request a review when the CI is green 🟢

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py [hunk 1]
```diff
@@ -599,6 +615,27 @@ def remove_special_characters(batch):
     # derive max & min input length for sample rate & max duration
     max_input_length = data_args.max_duration_in_seconds * feature_extractor.sampling_rate
     min_input_length = data_args.min_duration_in_seconds * feature_extractor.sampling_rate
+
+    if data_args.max_eval_duration_in_seconds is None:
+        logger.warning(
+            "`max_eval_duration_in_seconds` is not provided. Filtering audio samples in the eval set longer than"
```
Comment (patrickvonplaten) [inline examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:621]:
```suggestion
            f"`max_eval_duration_in_seconds` is not provided and `max_duration_in_seconds` is set to {max_duration_in_seconds}. Generally, it is not recommended to filter evaluation data, so we strongly recommend setting `max_eval_duration_in_seconds` either explicictely or setting it to `False`. For backwards compatibility,  audio samples in the eval set longer than `max_duration_in_seconds`: {max_duration_in_seconds} are filtered."
```

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py [hunk 2]
```diff
@@ -599,6 +615,27 @@ def remove_special_characters(batch):
     # derive max & min input length for sample rate & max duration
     max_input_length = data_args.max_duration_in_seconds * feature_extractor.sampling_rate
     min_input_length = data_args.min_duration_in_seconds * feature_extractor.sampling_rate
+
+    if data_args.max_eval_duration_in_seconds is None:
+        logger.warning(
```
Comment (patrickvonplaten) [inline examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:620]: For deprecation warnings we usually use `warnings.warn("Text", FutureWarning)` as shown here: https://github.com/huggingface/transformers/blob/957ce6465aa7f2cf89cbfd37d6cf166dd588ceab/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L1216

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py [hunk 3]
```diff
@@ -207,15 +207,31 @@ class DataTrainingArguments:
...
             "help": (
-                "Filter audio files that are longer than `max_duration_in_seconds` seconds to"
-                " 'max_duration_in_seconds`"
+                "Filter audio files in the evaluation set that are longer than `max_eval_duration_in_seconds` seconds"
             )
         },
     )
-    min_duration_in_seconds: float = field(
-        default=0.0, metadata={"help": "Filter audio files that are shorter than `min_duration_in_seconds` seconds"}
+    min_eval_duration_in_seconds: float = field(
```
Comment (patrickvonplaten) [inline examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:228]:
```suggestion
    min_eval_duration_in_seconds: Optional[Union[float, bool]] = field(
```
For backwards compatibility I'd leave it exactly like is, but also allow one to set it simply to `False` - see comment below to better understand

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py (hunk 1 above)
Comment (patrickvonplaten) [reply examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:621]: I think we should offer the user here an option to easily disable filtering

Comment (patrickvonplaten) [reply examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:621]: See comment above about allowing `max_eval_duration_in_seconds` to be a boolean set to False

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py [hunk 4]
```diff
@@ -599,6 +615,27 @@ def remove_special_characters(batch):
...
+            "`max_eval_duration_in_seconds` is not provided. Filtering audio samples in the eval set longer than"
+            " `max_duration_in_seconds`. From Transformers v5, the eval set will not be filtered by maximum audio"
+            " length if `max_eval_duration_in_seconds` is not provided."
+        )
+        max_eval_input_length = max_input_length
+    else:
+        max_eval_input_length = data_args.max_eval_duration_in_seconds * feature_extractor.sampling_rate
+
+    if data_args.min_eval_duration_in_seconds is None:
+        logger.warning(
```
Comment (patrickvonplaten) [inline examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:630]: Same comments as above - just no need to allow a boolean. But we should incentivize the reader to set `min_eval_duration_in_seconds` to 0

**Comment by sanchit-gandhi** [comment]
Thanks for the ping, will address in the coming days!

**Comment by patrickvonplaten** [comment]
Think this is worth fixing still, shouldn't take too long
